t/error_handler/error_handler.t

```diff
@@ -0,0 +1,214 @@
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 2: test apisix with internal error code 500
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 500
+--- error_code: 500
+--- response_body_like
+apisix.apache.org
+
+
+
+=== TEST 3: test apisix with internal error code 502
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 502
+--- error_code: 502
+--- response_body_like
+apisix.apache.org
+
+
+
+=== TEST 4: test apisix with internal error code 503
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 503
+--- error_code: 503
+--- response_body_like
+apisix.apache.org
+
+
+
+=== TEST 5: test apisix with internal error code 504
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 504
+--- error_code: 504
+--- response_body_like
+apisix.apache.org
+
+
+
+=== TEST 6: test apisix with internal error code 400
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 400
+--- error_code: 400
+--- response_body
+<html>
+<head><title>400 Bad Request</title></head>
+<body>
+<center><h1>400 Bad Request</h1></center>
+<hr><center> <a href="https://apisix.apache.org/">APISIX</a></center>
+</body>
+</html>
+
+
+
+=== TEST 7: test apisix with internal error code 401
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 401
+--- error_code: 401
+--- response_body
+<html>
+<head><title>401 Unauthorized</title></head>
+<body>
+<center><h1>401 Unauthorized</h1></center>
+<hr><center> <a href="https://apisix.apache.org/">APISIX</a></center>
+</body>
+</html>
+
+
+
+=== TEST 8: test apisix with internal error code 403
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 403
+--- error_code: 403
+--- response_body
+<html>
+<head><title>403 Forbidden</title></head>
+<body>
+<center><h1>403 Forbidden</h1></center>
+<hr><center> <a href="https://apisix.apache.org/">APISIX</a></center>
+</body>
+</html>
+
+
+
+=== TEST 9: test apisix with internal error code 404
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 404
+--- error_code: 404
+--- response_body
+<html>
+<head><title>404 Not Found</title></head>
+<body>
+<center><h1>404 Not Found</h1></center>
+<hr><center> <a href="https://apisix.apache.org/">APISIX</a></center>
+</body>
+</html>
+
+
+
+=== TEST 10: test apisix with internal error code 405
+--- request
+GET /hello
+--- more_headers
+X-Test-Status: 405
+--- error_code: 405
+--- response_body
+<html>
+<head><title>405 Not Allowed</title></head>
+<body>
+<center><h1>405 Not Allowed</h1></center>
+<hr><center> <a href="https://apisix.apache.org/">APISIX</a></center>
+</body>
+</html>
+
+
+
+=== TEST 12: test apisix with upstream error code 400
+--- request
+GET /specific_status
+--- more_headers
+X-Test-Upstream-Status: 400
+--- error_code: 400
+--- response_body
+upstream status: 400
+
+
+
+=== TEST 11: test apisix with upstream error code 500
+--- request
+GET /specific_status
+--- more_headers
+X-Test-Upstream-Status: 500
+--- error_code: 500
+--- response_body
+upstream status: 500
```

Comment:
I am a little confused, why do special treatments for upstream status?

Reply:
Here is a mock that I mock the upstream status.
You can see `t/lib/server.lua`.

Reply:
I mean, why doesn’t the upstream 5xx show the 5xx page of Apache APISIX?

Reply:
> I mean, why doesn’t the upstream 5xx show the 5xx page of Apache APISIX?

The upstream 5xx has its own mean.There maybe contains the error info like `{"code": 5xx, "error_msg": "xxx"}.
So IMO we shouldn't modify it.

Reply:
Do users need to care about the error of the Apache APISIX or the upstream error? I remember that there are already corresponding HTTP headers to distinguish

Reply:
In some situations. The info in the response body is useful. The user will need the upstream error info.
cc @tokers

Reply:
@moonming We should keep the upstream 4xx/5xx pages intact by default, but we may add a switch like:

```
proxy_intercept_errors on | off;
```

This is a standard Nginx directive.